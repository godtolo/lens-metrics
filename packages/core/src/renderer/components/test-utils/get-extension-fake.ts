/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { LensMainExtensionParts } from "../../../extensions/lens-main-extension";
import { LensMainExtension } from "../../../extensions/lens-main-extension";
import type { LensRendererExtensionParts } from "../../../extensions/lens-renderer-extension";
import { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import type { SetRequired } from "type-fest";
import type { LensExtensionManifest } from "../../../features/extensions/common/installed-extension";

export class TestExtensionMain extends LensMainExtension {}
export class TestExtensionRenderer extends LensRendererExtension {}

type FakeManifest = SetRequired<Partial<LensExtensionManifest>, "name">;

export interface FakeExtensionOptions {
  id: string;
  manifest: FakeManifest;
  rendererOptions?: Partial<LensRendererExtensionParts>;
  mainOptions?: Partial<LensMainExtensionParts>;
}

export const getExtensionFakeForMain = ({ id, manifest, mainOptions: options = {}}: FakeExtensionOptions) => (
  Object.assign(
    new TestExtensionMain({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: {
        version: "1.0.0",
        engines: {
          lens: "^5.5.0",
        },
        ...manifest,
      },
      manifestPath: "irrelevant",
    }),
    options,
  )
);

export const getExtensionFakeForRenderer = ({ id, manifest, rendererOptions: options = {}}: FakeExtensionOptions) => (
  Object.assign(
    new TestExtensionRenderer({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: {
        version: "1.0.0",
        engines: {
          lens: "^5.5.0",
        },
        ...manifest,
      },
      manifestPath: "irrelevant",
    }),
    options,
  )
);
