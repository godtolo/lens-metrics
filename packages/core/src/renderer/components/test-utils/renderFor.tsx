/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import React from "react";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { DiContainer } from "@ogre-tools/injectable";
import { DiContextProvider } from "@ogre-tools/injectable-react";

export type DiRender = (ui: React.ReactElement) => RenderResult;

type DiRenderFor = (di: DiContainer) => DiRender;

const getDOM = (di: DiContainer, ui: React.ReactElement) => (
  <DiContextProvider value={{ di }}>
    {ui}
  </DiContextProvider>
);

export const renderFor: DiRenderFor = (di) => (ui) => {
  const result = render(getDOM(di, ui));

  return {
    ...result,
    rerender: (ui) => result.rerender(getDOM(di, ui)),
  };
};
