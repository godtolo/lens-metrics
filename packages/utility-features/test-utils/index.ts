export * from "./src/flush-promises";
export * from "./src/get-global-override-for-function";
export * from "./src/get-global-override";
export * from "./src/get-promise-status";
