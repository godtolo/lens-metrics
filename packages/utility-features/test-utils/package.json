{
  "name": "@k8slens/test-utils",
  "version": "1.0.0-alpha.1",
  "description": "A collection of utilities for testing",
  "type": "commonjs",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "private": false,
  "files": [
    "dist"
  ],
  "author": {
    "name": "OpenLens Authors",
    "email": "info@k8slens.dev"
  },
  "license": "MIT",
  "homepage": "https://github.com/lensapp/lens",
  "scripts": {
    "build": "webpack",
    "dev": "webpack --mode=development --watch",
    "test": "jest --coverage --runInBand"
  }
}
