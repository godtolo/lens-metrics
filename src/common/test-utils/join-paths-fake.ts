/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { join } from "lodash/fp";
import type { JoinPaths } from "../path/join-paths.injectable";

export const joinPathsFake: JoinPaths = join("/");
