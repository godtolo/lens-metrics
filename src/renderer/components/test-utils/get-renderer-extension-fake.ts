/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { SetRequired } from "type-fest";
import { LensRendererExtension } from "../../../extensions/lens-renderer-extension";

export class TestExtension extends LensRendererExtension {}

export const getRendererExtensionFake = ({ id, ...rest }: SetRequired<Partial<LensRendererExtension>, "id">) => {
  const instance = new TestExtension({
    id,
    absolutePath: "irrelevant",
    isBundled: false,
    isCompatible: false,
    isEnabled: false,
    manifest: { name: id, version: "some-version" },
    manifestPath: "irrelevant",
  });

  Object.assign(instance, rest);

  return instance;
};
