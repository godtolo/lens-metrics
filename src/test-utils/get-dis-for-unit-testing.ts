/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getDiForUnitTesting as getRendererDi } from "../renderer/getDiForUnitTesting";
import { getDiForUnitTesting as getMainDi } from "../main/getDiForUnitTesting";
import { overrideIpcBridge } from "./override-ipc-bridge";
import { Volume } from "memfs";

export interface GetDisForUnitTestingOptions {
  doGeneralOverrides?: boolean;
}

export const getDisForUnitTesting = (opts: GetDisForUnitTestingOptions = {}) => {
  const memFsVolume = Volume.fromNestedJSON({
    "some-bin-directory": {},
    "some-directory-for-lens-local-storage": {},
    "some-electron-app-path-for-home": {},
    "some-electron-app-path-for-app-data": {},
    "some-electron-app-path-for-cache": {},
    "some-electron-app-path-for-temp": {},
    "some-electron-app-path-for-exe": {},
    "some-electron-app-path-for-module": {},
    "some-electron-app-path-for-desktop": {},
    "some-electron-app-path-for-documents": {},
    "some-electron-app-path-for-downloads": {},
    "some-electron-app-path-for-music": {},
    "some-electron-app-path-for-pictures": {},
    "some-electron-app-path-for-videos": {},
    "some-electron-app-path-for-logs": {},
    "some-electron-app-path-for-crash-dumps": {},
    "some-electron-app-path-for-crash-recent": {},
  }, "/");

  const rendererDi = getRendererDi({ ...opts, memFsVolume });
  const mainDi = getMainDi({ ...opts, memFsVolume });

  overrideIpcBridge({ rendererDi, mainDi });

  return {
    rendererDi,
    mainDi,
    runSetups: () => Promise.all([rendererDi.runSetups(), mainDi.runSetups()]),
  };
};
