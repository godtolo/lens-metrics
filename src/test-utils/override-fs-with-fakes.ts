/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import fsInjectable from "../common/fs/fs.injectable";
import { createFsFromVolume, Volume } from "memfs";
import type { readJsonSync as readJsonSyncImpl, writeJsonSync as writeJsonSyncImpl } from "fs-extra";

export const getOverrideFsWithFakes = () => {
  const root = createFsFromVolume(Volume.fromJSON({}));

  const readJsonSync = ((file, opts) => {
    const options = typeof opts === "string"
      ? {
        encoding: opts,
      }
      : opts;
    const value = root.readFileSync(file, options as any) as string;

    return JSON.parse(value, options?.reviver);
  }) as typeof readJsonSyncImpl;
  const writeJsonSync = ((file, object, opts) => {
    const options = typeof opts === "string"
      ? {
        encoding: opts,
      }
      : opts;

    root.writeFileSync(file, JSON.stringify(object, options?.replacer, options?.spaces), options as any);
  }) as typeof writeJsonSyncImpl;

  return (di: DiContainer) => {
    di.override(fsInjectable, () => ({
      pathExists: async (path) => root.existsSync(path),
      pathExistsSync: root.existsSync,
      readFile: root.promises.readFile as any,
      readFileSync: root.readFileSync as any,
      readJson: async (file, opts) => readJsonSync(file, opts),
      readJsonSync,
      writeFile: root.promises.writeFile as any,
      writeFileSync: root.writeFileSync as any,
      writeJson: async (file, obj, opts) => writeJsonSync(file, obj, opts as any),
      writeJsonSync,
      readdir: root.promises.readdir as any,
      lstat: root.promises.lstat as any,
      rm: root.promises.rm,
      access: root.promises.access,
    }));
  };
};
