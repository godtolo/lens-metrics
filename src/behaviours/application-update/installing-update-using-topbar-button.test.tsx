import type { AsyncFnMock } from "@async-fn/jest";
import asyncFn from "@async-fn/jest";
import type { RenderResult } from "@testing-library/react";
import type { CheckForPlatformUpdates } from "../../main/application-update/check-for-platform-updates/check-for-platform-updates.injectable";
import checkForPlatformUpdatesInjectable from "../../main/application-update/check-for-platform-updates/check-for-platform-updates.injectable";
import type { DownloadPlatformUpdate } from "../../main/application-update/download-platform-update/download-platform-update.injectable";
import downloadPlatformUpdateInjectable from "../../main/application-update/download-platform-update/download-platform-update.injectable";
import publishIsConfiguredInjectable from "../../main/application-update/publish-is-configured.injectable";
import periodicalCheckForUpdateWarningInjectable from "../../main/application-update/update-warning-level/periodical-check-for-update-warning.injectable";
import electronUpdaterIsActiveInjectable from "../../main/electron-app/features/electron-updater-is-active.injectable";
import type { ApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";
import { getApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";

describe("encourage user to update when sufficient time passed since update was downloaded", () => {
  let applicationBuilder: ApplicationBuilder;
  let checkForPlatformUpdatesMock: AsyncFnMock<CheckForPlatformUpdates>;
  let downloadPlatformUpdateMock: AsyncFnMock<DownloadPlatformUpdate>;

  beforeEach(() => {
    applicationBuilder = getApplicationBuilder();

    applicationBuilder.beforeApplicationStart(({ mainDi }) => {
      checkForPlatformUpdatesMock = asyncFn();
      downloadPlatformUpdateMock = asyncFn();

      mainDi.override(
        checkForPlatformUpdatesInjectable,
        () => checkForPlatformUpdatesMock,
      );

      mainDi.override(
        downloadPlatformUpdateInjectable,
        () => downloadPlatformUpdateMock,
      );

      mainDi.override(electronUpdaterIsActiveInjectable, () => true);
      mainDi.override(publishIsConfiguredInjectable, () => true);

      mainDi.unoverride(periodicalCheckForUpdateWarningInjectable);
      mainDi.permitSideEffects(periodicalCheckForUpdateWarningInjectable);
    })
  });

  describe("when started", () => {
    let rendered: RenderResult;

    beforeEach(async () => {
      rendered = await applicationBuilder.render();
    });

    it("renders", () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });

    it("does not show update button yet", () => {
      const button = rendered.queryByTestId("update-button");

      expect(button).toBeNull();
    });

    describe("given the update check", () => {
      let processCheckingForUpdatesPromise: Promise<void>;

      beforeEach(async () => {
        // TODO: initiate update check process automatically, not from tray
        processCheckingForUpdatesPromise = applicationBuilder.tray.click("check-for-updates");
      });

      describe("when update downloaded", () => {
        beforeEach(async () => {
          await checkForPlatformUpdatesMock.resolve({
            updateWasDiscovered: true,
            version: "some-version",
          });
          await downloadPlatformUpdateMock.resolve({ downloadWasSuccessful: true });
          await processCheckingForUpdatesPromise;
        });

        it("shows update button to help user to update", () => {
          const button = rendered.queryByTestId("update-button");
  
          expect(button).toBeInTheDocument();
        })
      });
    });
  });
});