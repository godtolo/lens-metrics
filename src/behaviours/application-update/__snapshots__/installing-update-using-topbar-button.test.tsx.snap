// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encourage user to update when sufficient time passed since update was downloaded when started renders 1`] = `
<body>
  <div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
    <div
      class="Notifications flex column align-flex-end"
    />
  </div>
</body>
`;
